include/kernels/laplace.cuh

```diff
@@ -29,27 +32,36 @@ namespace quda
 
     typedef typename mapper<Float>::type real;
 
-    F out;        /** output vector field */
-    const F in;   /** input vector field */
-    const F in_pack; /** input vector field used in packing to be able to independently resetGhost */
-    const F x;    /** input vector field for xpay*/
+    static constexpr unsigned int max_n_src = MAX_MULTI_RHS;
+    const int_fastdiv n_src;
+
+    F out[max_n_src];        /** output vector field */
+    F in[max_n_src];   /** input vector field */
+    const Ghost halo_pack;   /** accessor used for writing the halo field */
+    const Ghost halo;        /** accessor used for reading the halo field */
+    F x[max_n_src];    /** input vector field for xpay*/
     const G U;    /** the gauge field */
     const real a; /** xpay scale factor - can be -kappa or -kappa^2 */
     const real b; /** used by Wuppetal smearing kernel */
     int dir;      /** The direction from which to omit the derivative */
 
-    LaplaceArg(ColorSpinorField &out, const ColorSpinorField &in, const GaugeField &U, int dir, double a, double b,
-               const ColorSpinorField &x, int parity, bool dagger, const int *comm_override) :
-      DslashArg<Float, nDim>(out, in, U, x, parity, dagger, a != 0.0 ? true : false, 1, false, comm_override),
-      out(out),
-      in(in),
-      in_pack(in),
-      x(x),
+    LaplaceArg(cvector_ref<ColorSpinorField> &out, cvector_ref<const ColorSpinorField> &in, const ColorSpinorField &halo,
+               const GaugeField &U, int dir, double a, double b, cvector_ref<const ColorSpinorField> &x,
+               int parity, bool dagger, const int *comm_override) :
+      DslashArg<Float, nDim>(out, in, halo, U, x, parity, dagger, a != 0.0 ? true : false, 1, false, comm_override),
+      halo_pack(halo),
+      halo(halo),
       U(U),
       a(a),
       b(b),
       dir(dir)
     {
+      if (out.size() > max_n_src) errorQuda("vector set size %lu greater than max size %d", out.size(), max_n_src);
```

Comment:
This check should be in more files (or we should eliminate it from here and guarantee it elsewhere)

Reply:
This aspect will be addressed with this commit https://github.com/lattice/quda/pull/1465/commits/314468707963b1b30cbe80204957e62f23d725c4, where the `instantiate` function used for the Dslash functions will deal with auto splitting the fields as needed.  With this done, we can safely remove all these checks.  I'll leave a single check in the `DslashArg` constructor though, since that will guarantee it is called by every operator.

Reply:
e679e8216 puts the check in `DslashArg` and removes it from all child operators.